scrapy/crawler.py

```diff
@@ -72,9 +72,11 @@ def crawl(self, *args, **kwargs):
             start_requests = iter(self.spider.start_requests())
             yield self.engine.open_spider(self.spider, start_requests)
             yield defer.maybeDeferred(self.engine.start)
-        except Exception:
+        except Exception as e:
             self.crawling = False
-            raise
+            if self.engine is not None:
+                yield self.engine.close()
+            raise e
```

Comment:
Hm, didn't catch that `raise e` doesn't preserve the traceback :/ The problem is that lines 77/78 somehow make Python forget the last active exception, and therefore `raise` gives a `TypeError` because I'm not allowed to `raise None` (and it wouldn't be what we want anyways). I guess the forgetting happens somewhere inside twisted's `yield` magic, but no idea.

`raise e` does preserve the traceback on python3, and there's an ugly workaround for python2 (4th code block [here](http://www.ianbicking.org/blog/2007/09/re-raising-exceptions.html)). Maybe that will do inside an `if six.PY2` block?